### ISSUE TITLE
sql: support ALTER TABLE ... ALTER COLUMN ... SET DATA TYPE ...

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#9851]
Requested by @jayesh_nambiar_twitter on gitter, suggested by @dt:

you can add the new col, use an `UPDATE tbl SET newcol = f(oldcol)` to move the data over, then drop the old col
(and then rename `newcol` to `oldcol`)
`f` there could be as simple as a `::newtype` cast, or some other builtin function that computes the new values from the old ones

**Comment by bdarnell** [comment]
See also #5950, which asks for a simpler subset of type changes (changing the length limits of string types, needed by the django ORM).

**Comment by yznming** [comment]
@knz  Can I have a try?

**Comment by knz** [comment]
Hi @yznming yes you can try this, but this is not my area of expertise so I am not sure I can help you.

**Comment by yznming** [comment]
@knz  you have gave me lots of help, thank you so much :)

**Comment by jordanlewis** [comment]
Based on an informal survey of popular open source Ruby on Rails apps, it seems like this is a very common operation - developers often add migrations that change column types as their apps evolved.

We should make an effort to support this in the not-too-distant future.

**Comment by mclark4386** [comment]
Yeah, this is a big deal in almost every framework I've ever used... I'm currently trying to implement a workaround for the [Buffalo](https://gobuffalo.io) framework for go, but have run into a [weird snag](https://github.com/cockroachdb/cockroach/issues/21032) if anyone could help that would be at least one platform that could support CockroachDB out of the box with all migrations supported.

**Comment by knz** [comment]
cc @awoods187 this feature is coming up more and more for compatibility with other tools, perhaps it is time to re-prioritize.

**Comment by bobvawter** [comment]
Drafting an RFC here https://github.com/cockroachdb/cockroach/pull/24703

**Comment by tim-o** [comment]
Zendesk ticket #[2497](https://cockroachdb.zendesk.com/agent/tickets/2497) has been linked to this issue.

**Comment by awoods187** [comment]
top requested feature based on telemetry data

**Comment by tim-o** [comment]
*Comment made from Zendesk by Ricardo Rocha on 2019-07-08 at 14:23:*

<blockquote><div class="zd-comment" dir="auto"><p dir="auto">Closing older ZenDesk forum incidents as per Support meeting on 7/2</p></div></blockquote>

**Comment by markelog** [comment]
Why this was closed?

**Comment by jordanlewis** [comment]
Good question. It shouldn't have been.

**Comment by tim-o** [comment]
That was our error, apologies.

**Comment by tim-o** [comment]
*Comment made from Zendesk by Ricardo Rocha on 2019-07-15 at 17:26:*

<blockquote><div class="zd-comment" dir="auto"><p dir="auto">Reopening since bulk update closed incorrectly.</p></div></blockquote>

**Comment by markelog** [comment]
Cool, thanks :), excited about the possibility of this feature btw

**Comment by sandstrom** [comment]
@RichardJCai In which commit was this fixed? (just curious, I can't find it)

**Comment by Luit** [comment]
Apparently, not yet fully in 20.2.0-alpha.1: 

    ERROR:  internal error: unknown conversion for VARCHAR(20) -> VARCHAR(10)

**Comment by knz** [comment]
cc @jordanlewis for re-triage

**Comment by RichardJCai** [comment]
> Apparently, not yet fully in 20.2.0-alpha.1:
> 
> ```
> ERROR:  internal error: unknown conversion for VARCHAR(20) -> VARCHAR(10)
> ```

This should be fixed in 20.2.0 alpha 2.

**Comment by RichardJCai** [comment]
> @RichardJCai In which commit was this fixed? (just curious, I can't find it)

I think the PR automatically closed the ticket.

Here is the commit that experimental support for ALTER COLUMN TYPE where the column is not part of an index. There is still some work left to be done for cases where the column is part of an index, has a constraint, and it's currently not supported inside a transaction.

https://github.com/cockroachdb/cockroach/pull/46933/commits/dcd242935afbd3b00d59bc279729f4a2bdda6982

**Comment by jordanlewis** [comment]
Hey @solongordon can we close this in favor of #49329 at this point, or is there still more work to be done that isn't a part of the `enable_experimental_alter_column_type_general` flag?
